File: src/pathtracer/blue-noise.ts
/**
 * Generate a tileable blue noise texture using the void-and-cluster algorithm
 * (Ulichney 1993, with sigma=1.9 per demofox recommendations).
 *
 * The algorithm produces a progressive blue noise pattern: thresholding the
 * output at ANY value produces a well-distributed point set. This is critical
 * for path tracer denoising — the spatial filter sees uniform noise at every scale.
 *
 * Three phases:
 *   1. Create + stabilize an initial binary pattern (~10% seed density)
 *   2. Rank seed points (remove tightest clusters) + fill to 50% (add at largest voids)
 *   3. Fill 50%→100% with inverted energy (track 0-pixels, remove tightest 0-clusters)
 *
 * Returns Float32Array of size×size values in [0, 1).
 */
export function generateBlueNoise(size: number = 128): Float32Array {
    const N = size * size;
    const sigma = 1.9;
    const sigma2x2 = 2.0 * sigma * sigma;
    const radius = Math.ceil(3.0 * sigma); // 6 pixels

    const binary = new Uint8Array(N);
    const energy = new Float32Array(N);
    const ranks = new Float32Array(N);

    // ── Helpers ──────────────────────────────────────────────────────────

    // Add or subtract a pixel's Gaussian contribution to the energy LUT
    function updateEnergy(idx: number, sign: number): void {
        const cx = idx % size;
        const cy = (idx / size) | 0;
        for (let dy = -radius; dy <= radius; dy++) {
            for (let dx = -radius; dx <= radius; dx++) {
                const nx = ((cx + dx) % size + size) % size;
                const ny = ((cy + dy) % size + size) % size;
                energy[ny * size + nx] += sign * Math.exp(-(dx * dx + dy * dy) / sigma2x2);
            }
        }
    }

    // Find pixel with binary[i]==val that has max or min energy
    function findExtreme(val: number, findMax: boolean): number {
        let bestIdx = 0;
        let bestE = findMax ? -Infinity : Infinity;
        for (let i = 0; i < N; i++) {
            if (binary[i] !== val) continue;
            if (findMax ? energy[i] > bestE : energy[i] < bestE) {
                bestE = energy[i];
                bestIdx = i;
            }
        }
        return bestIdx;
    }

    // Simple xorshift32 RNG
    let seed = 2463534242;
    const rng = (): number => {
        seed ^= seed << 13;
        seed ^= seed >> 17;
        seed ^= seed << 5;
        return (seed >>> 0) / 4294967296;
    };

    // ── Phase 1: Create initial binary pattern ──────────────────────────
    // Seed ~10% of pixels randomly
    const initialCount = Math.max(1, Math.floor(N * 0.1));
    let onesCount = 0;
    while (onesCount < initialCount) {
        const idx = Math.floor(rng() * N);
        if (binary[idx]) continue;
        binary[idx] = 1;
        updateEnergy(idx, 1);
        onesCount++;
    }

    // Stabilize: repeatedly move tightest cluster → largest void until convergent
    for (let iter = 0; iter < N; iter++) {
        const clusterIdx = findExtreme(1, true); // 1-pixel with max energy
        binary[clusterIdx] = 0;
        updateEnergy(clusterIdx, -1);

        const voidIdx = findExtreme(0, false); // 0-pixel with min energy
        if (voidIdx === clusterIdx) {
            // Converged: re-insert at same position and stop
            binary[clusterIdx] = 1;
            updateEnergy(clusterIdx, 1);
            break;
        }

        binary[voidIdx] = 1;
        updateEnergy(voidIdx, 1);
    }

    // ── Rank initial seed points (remove tightest clusters) ─────────────
    // Save the stabilized initial pattern
    const savedBinary = new Uint8Array(binary);
    const savedEnergy = new Float32Array(energy);
    let remaining = onesCount;

    while (remaining > 0) {
        const clusterIdx = findExtreme(1, true);
        remaining--;
        ranks[clusterIdx] = remaining; // most clustered → highest rank among seeds
        binary[clusterIdx] = 0;
        updateEnergy(clusterIdx, -1);
    }

    // ── Phase 2: Fill from initial pattern to 50% density ───────────────
    // Restore initial binary pattern
    binary.set(savedBinary);
    energy.set(savedEnergy);
    let currentRank = onesCount;
    const halfN = Math.floor(N / 2);

    while (currentRank < halfN) {
        const voidIdx = findExtreme(0, false); // 0-pixel with min energy = largest void
        ranks[voidIdx] = currentRank;
        binary[voidIdx] = 1;
        updateEnergy(voidIdx, 1);
        currentRank++;
    }

    // ── Phase 3: Fill from 50% to 100% (inverted energy) ────────────────
    // Rebuild energy from 0-pixels (invert meaning of 0s and 1s)
    energy.fill(0);
    for (let i = 0; i < N; i++) {
        if (binary[i] === 0) {
            updateEnergy(i, 1);
        }
    }

    while (currentRank < N) {
        // Tightest cluster of 0s = 0-pixel with max energy
        const clusterIdx = findExtreme(0, true);
        ranks[clusterIdx] = currentRank;
        binary[clusterIdx] = 1;
        updateEnergy(clusterIdx, -1); // remove its 0-contribution
        currentRank++;
    }

    // Normalize ranks to [0, 1)
    for (let i = 0; i < N; i++) {
        ranks[i] /= N;
    }

    return ranks;
}

File: rust/kansei-core/src/pathtracer/blue_noise.rs
/// Blue noise table dimensions.
pub const BLUE_NOISE_SIZE: u32 = 128;

/// Generate a tileable blue noise texture using the void-and-cluster algorithm
/// (Ulichney 1993, sigma=1.9). Produces a progressive blue noise pattern where
/// thresholding at any value gives a well-distributed point set.
///
/// Returns `128 * 128 = 16384` float values in [0, 1).
pub fn generate_blue_noise() -> Vec<f32> {
    let size = BLUE_NOISE_SIZE as usize;
    let n = size * size;
    let sigma: f32 = 1.9;
    let sigma2x2 = 2.0 * sigma * sigma;
    let radius = (3.0 * sigma).ceil() as i32;

    let mut binary = vec![0u8; n];
    let mut energy = vec![0.0f32; n];
    let mut ranks = vec![0.0f32; n];

    // xorshift32 RNG
    let mut seed: u32 = 2463534242;
    let mut rng = || -> f32 {
        seed ^= seed << 13;
        seed ^= seed >> 17;
        seed ^= seed << 5;
        (seed as f32) / 4294967296.0
    };

    // Add/subtract a pixel's Gaussian contribution to energy
    let update_energy =
        |energy: &mut [f32], idx: usize, sign: f32| {
            let cx = (idx % size) as i32;
            let cy = (idx / size) as i32;
            for dy in -radius..=radius {
                for dx in -radius..=radius {
                    let nx = ((cx + dx) % size as i32 + size as i32) as usize % size;
                    let ny = ((cy + dy) % size as i32 + size as i32) as usize % size;
                    let d2 = (dx * dx + dy * dy) as f32;
                    energy[ny * size + nx] += sign * (-d2 / sigma2x2).exp();
                }
            }
        };

    // Find pixel with binary[i]==val that has max (find_max=true) or min energy
    let find_extreme = |binary: &[u8], energy: &[f32], val: u8, find_max: bool| -> usize {
        let mut best_idx = 0;
        let mut best_e = if find_max { f32::NEG_INFINITY } else { f32::INFINITY };
        for i in 0..n {
            if binary[i] != val {
                continue;
            }
            if (find_max && energy[i] > best_e) || (!find_max && energy[i] < best_e) {
                best_e = energy[i];
                best_idx = i;
            }
        }
        best_idx
    };

    // Phase 1: Create initial binary pattern (~10% seed density)
    let initial_count = (n as f32 * 0.1).max(1.0) as usize;
    let mut ones_count = 0;
    while ones_count < initial_count {
        let idx = (rng() * n as f32) as usize % n;
        if binary[idx] != 0 {
            continue;
        }
        binary[idx] = 1;
        update_energy(&mut energy, idx, 1.0);
        ones_count += 1;
    }

    // Stabilize: move tightest cluster → largest void until convergent
    for _ in 0..n {
        let cluster_idx = find_extreme(&binary, &energy, 1, true);
        binary[cluster_idx] = 0;
        update_energy(&mut energy, cluster_idx, -1.0);

        let void_idx = find_extreme(&binary, &energy, 0, false);
        if void_idx == cluster_idx {
            binary[cluster_idx] = 1;
            update_energy(&mut energy, cluster_idx, 1.0);
            break;
        }

        binary[void_idx] = 1;
        update_energy(&mut energy, void_idx, 1.0);
    }

    // Rank initial seed points (remove tightest clusters)
    let saved_binary = binary.clone();
    let saved_energy = energy.clone();
    let mut remaining = ones_count;

    while remaining > 0 {
        let cluster_idx = find_extreme(&binary, &energy, 1, true);
        remaining -= 1;
        ranks[cluster_idx] = remaining as f32;
        binary[cluster_idx] = 0;
        update_energy(&mut energy, cluster_idx, -1.0);
    }

    // Phase 2: Fill from initial pattern to 50%
    binary.copy_from_slice(&saved_binary);
    energy.copy_from_slice(&saved_energy);
    let mut current_rank = ones_count;
    let half_n = n / 2;

    while current_rank < half_n {
        let void_idx = find_extreme(&binary, &energy, 0, false);
        ranks[void_idx] = current_rank as f32;
        binary[void_idx] = 1;
        update_energy(&mut energy, void_idx, 1.0);
        current_rank += 1;
    }

    // Phase 3: Fill from 50% to 100% (inverted energy)
    energy.fill(0.0);
    for i in 0..n {
        if binary[i] == 0 {
            update_energy(&mut energy, i, 1.0);
        }
    }

    while current_rank < n {
        let cluster_idx = find_extreme(&binary, &energy, 0, true);
        ranks[cluster_idx] = current_rank as f32;
        binary[cluster_idx] = 1;
        update_energy(&mut energy, cluster_idx, -1.0);
        current_rank += 1;
    }

    // Normalize to [0, 1)
    let n_f = n as f32;
    for r in &mut ranks {
        *r /= n_f;
    }

    ranks
}
